addons/html_editor/static/src/others/embedded_components/plugins/caption_plugin/caption_plugin.js

```diff
@@ -0,0 +1,233 @@
+            image.after(caption);
+        }
+    }
+
+    hasImageCaption(image) {
+        if (!image) {
+            return;
+        }
+        const block = closestBlock(image);
+        return (
+            block.nodeName === "FIGURE" && !!block.querySelector("[data-embedded='caption'] input")
+        );
+    }
+
+    toggleImageCaption(image) {
+        image = image || this.dependencies.image.getSelectedImage();
+        if (!image) {
+            return;
+        }
+        if (this.hasImageCaption(image)) {
+            this.removeImageCaption(image);
+        } else {
+            this.addImageCaption(image, image.getAttribute("data-caption") || "");
+        }
+    }
+
+    getCaptionId() {
+        return "" + Math.floor(Math.random() * Date.now());
+    }
+
+    addImageCaption(image, captionText = "", focusInput = true) {
+        image = image || this.dependencies.image.getSelectedImage();
+        if (!image) {
+            return;
+        }
+        // Move the image within a figure element.
+        const figure = this.document.createElement("figure");
+        const link = image.parentElement.nodeName === "A" && image.parentElement;
+        if (link && (link.previousSibling || link.nextSibling)) {
+            // <p>wx<a><img/></a>yz</p> => <p>wx</p><p><a><img/></a></p><p>yz</p>
+            this.dependencies.split.splitAroundUntil(link, closestBlock(link));
+        } else if (
+            !link &&
+            (image.previousSibling || image.nextSibling) &&
+            closestBlock(image) !== this.editable
+        ) {
+            // <p>wx<img/>yz</p> => <p>wx</p><p><img/></p><p>yz</p>
+            this.dependencies.split.splitAroundUntil(image, closestBlock(image));
+        }
+        // => <p><figure><img/></figure></p>
+        // or <p><a><figure><img/></figure></a></p>
+        image.before(figure);
+        figure.append(image);
+        if (!link && figure.parentElement !== this.editable) {
+            // => <figure><img/></figure></p>
+            // but still <p><a><figure><img/></figure></p>
+            unwrapContents(figure.parentElement);
+        }
+        // Set the caption and its ID.
+        const captionId = this.getCaptionId();
+        image.setAttribute("data-caption-id", captionId);
+        image.setAttribute("data-caption", captionText || "");
+        // Add the caption component.
+        // => <p><figure><img/><figcaption>...</figcaption></figure></p>
+        // or <p><a><figure><img/><figcaption>...</figcaption></figure></a></p>
+        const caption = renderToElement("html_editor.EmbeddedCaptionBlueprint", {
+            embeddedProps: JSON.stringify({
+                id: captionId,
+                focusInput,
+            }),
+        });
+        figure.append(caption);
+        // Ensure it's not possible to write inside the figure.
+        figure.setAttribute("contenteditable", "false");
+        // Ensure it's possible to write before and after the figure.
```

Comment:
should this be aware of the current editor `baseContainer`and use it, instead of always a `p` ?